CPP20.md

```diff
@@ -0,0 +1,274 @@
+```
+
+### Range-based for statements with initializer
+A new versions of the range-based for statement for C++: `for (init; decl : expr)`  
+which simplifies common code patterns, can help users keep scopes tight and offers an elegant solution to a common lifetime problem.
+```c++
+for (T thing = f(); auto& x : thing.items()) {
+    mutate(&x);
+    log(x);
+}
+
+for (std::size_t i = 0; const auto& x : foo()) {
+    bar(x, i++);
+}
+```
+
+## C++20 Library Features
+
+### Concepts library
+The concepts library provides definitions of fundamental library concepts that can be used to perform
+compile-time validation of template arguments and perform function dispatch based on properties of types.
+These concepts provide a foundation for equational reasoning in programs.
+
+Core language concepts
+```c++
+template <typename T, typename U>
+concept Common = std::Same<std::common_type_t<T, U>, std::common_type_t<U, T>> &&
+  requires {
+    static_cast<std::common_type_t<T, U>>(std::declval<T>());
+    static_cast<std::common_type_t<T, U>>(std::declval<U>());
+  } &&
+  std::CommonReference<std::add_lvalue_reference_t<const T>, std::add_lvalue_reference_t<const U>> &&
+  std::CommonReference<std::add_lvalue_reference_t<std::common_type_t<T, U>>,
+  std::common_reference_t<std::add_lvalue_reference_t<const T>, std::add_lvalue_reference_t<const U>>>;
+```
+
+Comparison concepts
+```c++
+template <typename B>
+concept Boolean = std::Movable<std::remove_cvref_t<B>> &&
+    requires(const std::remove_reference_t<B>& b1, const std::remove_reference_t<B>& b2, const bool a)
+    {
+        requires std::ConvertibleTo<const std::remove_reference_t<B>&, bool>;
+        !b1;      requires std::ConvertibleTo<decltype(!b1), bool>;
+        b1 && a;  requires std::Same<decltype(b1 && a), bool>;
+        b1 || a;  requires std::Same<decltype(b1 || a), bool>;
+        b1 && b2; requires std::Same<decltype(b1 && b2), bool>;
+        a && b2;  requires std::Same<decltype(a && b2), bool>;
+        b1 || b2; requires std::Same<decltype(b1 || b2), bool>;
+        a || b2;  requires std::Same<decltype(a || b2), bool>;
+        b1 == b2; requires std::ConvertibleTo<decltype(b1 == b2), bool>;
+        b1 == a;  requires std::ConvertibleTo<decltype(b1 == a), bool>;
+        a == b2;  requires std::ConvertibleTo<decltype(a == b2), bool>;
+        b1 != b2; requires std::ConvertibleTo<decltype(b1 != b2), bool>;
+        b1 != a;  requires std::ConvertibleTo<decltype(b1 != a), bool>;
+        a != b2;  requires std::ConvertibleTo<decltype(a != b2), bool>;
+  };
+```
+
+Object concepts
+```c++
+template <typename T>
+concept Movable = std::is_object_v<T> && std::MoveConstructible<T> &&
+                  std::Assignable<T&, T> && std::Swappable<T>;
+```
+
+Callable concepts
+```c++
+template <typename R, typename T, typename U>
+concept Relation = std::Predicate<R, T, T> && std::Predicate<R, U, U> &&
+                   std::Predicate<R, T, U> && std::Predicate<R, U, T>;
+```
```

Comment:
Instead of listing out code for some of these library concepts, it would probably be more useful to list most of them (i.e. the more useful ones) with a short description. Like so:

### Core language concepts
- **Same** - specifies two types are the same
- **DerivedFrom** - specifies a type is derived from another type
- etc...

### Comparison concepts
- etc...